### ISSUE TITLE
Why not using DSRL-NA?

### ISSUE THREAD

**Issue by hiyyg** [nakamotoo/dsrl_pi0#8]
Why is DSRL-SAC used for π₀ instead of DSRL-NA? I noticed that the paper does not discuss this — could the authors please provide some explanation?

**Comment by mw9385** [comment]
Hi, I have tried to implement the code and it needs much time to update the model. When you take a look at the Algorithm 1 in the paper, next-action and actions from random Gaussian noise must be sampled from diffusion policy (i.e., open-pi model). These process takes much longer time than SAC method. If you anyhow figure out a method that accelerates the inferring process, you could give a try.

**Comment by lucanunz** [comment]
Hi! You can find an implementation of the DSRL-NA version on our repo at https://github.com/arayabrain/dsrl_pi0_na .

It is still slower than the SAC version because of what @mw9385 said, but runs in feasible time at least (depending on the hardware you have).

These are the main changes that made it run in a feasible time frame
1) Save the KV cache during rollouts and reuse it during training steps. No more redundant computations for the same observations
2) Allow the VLA to infer over a batch instead of single input
3) Shard the VLA inference over multiple GPUs (if available)

This is how it compares with the baseline implemented on this repo (aloha cube env).
Note that for the baseline we could use a batch size of 256, while for the NA version we used a batch size of 128.

<img width="2528" height="1328" alt="Image" src="https://github.com/user-attachments/assets/94b01311-4636-4f53-81b3-4228037a1649" />

**Comment by mw9385** [comment]
Hi @lucanunz. Thanks you for your great works! I have some minor question regrading DSRL_NA implementation. In the original code, the authors tried to repeat the same noise as it has large action space. Did you also follow the same method or tried with original noise action space (i.e., chunk_size * noise_dim)?

**Comment by shivakanthsujit** [comment]
So we noticed the following
- For an undertrained model (<10k gradient steps), repeated noise did not work and you needed to use the whole noise space
- Once the model was trained to around 30k steps, the repeated noise worked well.

For these experiments with DSRL-NA, we followed the paper in doing repeated noise across chunks. We did not experiment with using the whole noise space. 

Hope this helps!

**Comment by mw9385** [comment]
@lucanunz Thanks. One more minor question regrading your code. 

How long it will take to do the one gradient step? Could you tell me the whole training time in wall time? If it works in reasonable time, I want to try it.

**Comment by lucanunz** [comment]
The training time depends on the setting of the run! But I will give the details on what we used and what you can find as default in our repo.

The parameters that matter for training time are
- Batch size
- Update to data (UTD) ratio and number of updates for each model (noise critic, action critic, noise actor)
- How many GPUs you have available

Our experiments are using
- Batch size of 128
- A UTD ratio of 20 (same as the baseline, so that the plot above is fair since we have gradient updates on the x axis), and we set the ratio of updates for the models as
    - 15 updates to the action critic, 5 updates to the noise critic and 5 updates for the noise actor
- Inference is sharded over 4 A100s

With these setting, we achieve around **2.4 - 2.7 iterations/second**.

You can experiment with lower UTDs and different updates per each model to save some time

**Comment by Grange007** [comment]
Hi @lucanunz thank you for your great work! May I ask if your dsrl_pi0_na supports pi0.5, or just pi0 now? I noticed you are using the newer version of openpi code which contains pi0.5.

**Comment by lucanunz** [comment]
Hi @Grange007 ! It does support pi0.5! The code we pushed for the `AlohaTransferCube-v0` environment loads the pretrained `pi0_aloha_sim` config which is a pi0 policy, but we did a lot of in-house experiments on our own datasets with pi0.5.
I believe the code changes should be minimal, if any, to go from pi0 to pi0.5. To adapt the code to your own environment most of the changes will be on handling the observations and environment.

**Comment by Grange007** [comment]
Wow thank you! Really awesome work! May I ask another thing, have you tried to deployed this dsrl_pi0_na on a real robot? I noticed this DSRL-NA is indeed much slower than DSRL-SAC, and noticed you added a lot of acceleration tricks to that. Will this also be acceptable for real-world scenarios where unlike just submitting a job in simulation, we have to watch the robot rollout a trajectory and then wait for the training?

**Comment by lucanunz** [comment]
Great question! We haven’t deployed this on a real robot yet, but the training speed should be acceptable also for real-world setups.

The waiting time with the hardware setup and settings I mentioned before, these are some rough timings:

- Initialization: You’ll have a one-time wait for the first training after initializing your buffer with N trajectories. In our code, after filling your buffer you will do at most 3000 updates (see [here](https://github.com/arayabrain/dsrl_pi0_na/blob/afcb58f2bd95a2e3771db359b8f98eb8f4f9ebee/examples/train_utils_aloha_sim_na.py#L73), which equals a training time of ~20min).

- Per-Rollout Training: After the initial phase, you’ll typically wait a few minutes for training between trajectories. This varies on the UTD you choose and how long are your episodes. Usually, these few minutes overlap with the time needed to manually reset your real robot environment. For example, if your episode is long 300 steps and your chunk size is 30, you will do `10*UTD` gradient steps. With a speed of `2.5it/s`, you will have to wait `4*UTD` seconds.

While DSRL-NA is computationally slower than the SAC version, it is more sample efficient so you'll need fewer total physical rollouts to have the same performance improvement. However, we haven't benchmarked whether the reduced data collection time fully offsets the increased training time in a real-world setting.

**Comment by Grange007** [comment]
Cool thank you! This really helped a lot. I'm now thinking if it would be possible to incorporate the asynchronous running and acting from [`hil-serl`](https://github.com/rail-berkeley/hil-serl) to dsrl. Then we can have the learning node running asynchronously when we are performing a rollout, and we won't need to wait for the updating between rollouts, which should improve the training efficiency. I'll try to implement this when I have time, which may be a solution to make DSRL-NA more acceptable. Also I'm thinking about if we can use much fewer flow matching steps during distillation, like we use 2 steps instead of original 10 steps? As the iterative flow matching should be the main bottleneck for each forward pass time. Maybe we can do some experiments to see how it works, not knowing if this will harm the performance.

**Comment by shivakanthsujit** [comment]
Yeah those are the directions we are exploring too. We used 5 flow steps for some experiments and didn't notice a degradation in performance. So lowering the flow steps would be a good way to improve the training speed/sample efficiency.

We'd done some testing to check how much the flow steps affects the inference speed and that's the table below. This is a local policy with batch size of 128, on a single A100, for the aloha_sim env. We compared with and without using the kv cache. In the actual training loop there are other overheads wrt moving the cache to the correct device, gathering the sharded actions across devices, so you might not see exactly these improvements, but it is a good rule-of-thumb to have. 

The inference speed doesnt change linearly with flow steps, and there is a trade-off with action quality, so 5 might be a good compromise. 

## Without kv_cache

| Flow step change | Avg total time change (ms) | Speedup |
|---|---:|---:|
| 10 → 5 | 2965.59 → 2768.07 (**-197.52 ms**) | **1.07× faster** |
| 5 → 2 | 2768.07 → 2669.98 (**-98.09 ms**) | **1.04× faster** |
| 10 → 2 | 2965.59 → 2669.98 (**-295.61 ms**) | **1.11× faster** |

## With kv_cache

| Flow step change | Avg total time change (ms) | Speedup |
|---|---:|---:|
| 10 → 5 | 459.90 → 262.14 (**-197.76 ms**) | **1.75× faster** |
| 5 → 2 | 262.14 → 155.78 (**-106.36 ms**) | **1.68× faster** |
| 10 → 2 | 459.90 → 155.78 (**-304.12 ms**) | **2.95× faster** |

**Comment by Grange007** [comment]
Wow you are really awesome guys, thank so much for your great job! That's really important findings!
